File: webrtc/src/main/java/dev/onvoid/webrtc/RTCIceServer.java
/*
 * Copyright 2019 Alex Andres
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package dev.onvoid.webrtc;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * The RTCIceServer is used to describe the STUN and TURN servers that can be
 * used by the ICE Agent to establish a connection with a peer.
 *
 * @author Alex Andres
 */
public class RTCIceServer {

	/**
	 * List of STUN or TURN URI(s) associated with this server. The "host" part
	 * of the URI may contain either an IP address or a hostname. Valid formats
	 * are described in RFC7064 and RFC7065.
	 */
	public List<String> urls;

	/**
	 * The username to be used when the TURN server requests authorization.
	 */
	public String username;

	/**
	 * The password to be used when the TURN server requests authorization.
	 */
	public String password;

	/**
	 * The TLS certificate policy.
	 */
	public TlsCertPolicy tlsCertPolicy;

	/**
	 * If the URIs in {@link #urls} only contain IP addresses, this field can be
	 * used to indicate the hostname, which may be necessary for TLS (using the
	 * SNI extension). If {@link #urls} itself contains the hostname, this isn't
	 * necessary.
	 */
	public String hostname;

	/**
	 * List of protocols to be used in the TLS ALPN extension.
	 */
	public List<String> tlsAlpnProtocols;

	/**
	 * List of elliptic curves to be used in the TLS elliptic curves extension.
	 * Only curve names supported by OpenSSL should be used (eg.
	 * "P-256","X25519").
	 */
	public List<String> tlsEllipticCurves;


	/**
	 * Creates an instance of RTCIceServer.
	 */
	public RTCIceServer() {
		urls = new ArrayList<>();
		tlsAlpnProtocols = new ArrayList<>();
		tlsEllipticCurves = new ArrayList<>();
		tlsCertPolicy = TlsCertPolicy.SECURE;
	}

	@Override
	public int hashCode() {
		return Objects.hash(hostname, password, tlsAlpnProtocols, tlsCertPolicy,
				tlsEllipticCurves, urls, username);
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj) {
			return true;
		}
		if (obj == null) {
			return false;
		}
		if (getClass() != obj.getClass()) {
			return false;
		}
		
		RTCIceServer other = (RTCIceServer) obj;
		
		return Objects.equals(hostname, other.hostname) &&
				Objects.equals(password, other.password) &&
				Objects.equals(tlsAlpnProtocols, other.tlsAlpnProtocols) &&
				tlsCertPolicy == other.tlsCertPolicy &&
				Objects.equals(tlsEllipticCurves, other.tlsEllipticCurves) &&
				Objects.equals(urls, other.urls) &&
				Objects.equals(username, other.username);
	}

	@Override
	public String toString() {
		return String.format("%s@%d [urls=%s, username=%s, password=%s, tlsCertPolicy=%s, hostname=%s, tlsAlpnProtocols=%s, tlsEllipticCurves=%s]",
						RTCIceServer.class.getSimpleName(), hashCode(),
						urls, username, password, tlsCertPolicy, hostname,
						tlsAlpnProtocols, tlsEllipticCurves);
	}
}

File: webrtc-jni/src/api/RTCIceServer.cpp
/*
 * Copyright 2019 Alex Andres
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "api/RTCIceServer.h"
#include "JavaArrayList.h"
#include "JavaClasses.h"
#include "JavaEnums.h"
#include "JavaIterable.h"
#include "JavaList.h"
#include "JavaRef.h"
#include "JavaObject.h"
#include "JavaString.h"
#include "JavaUtils.h"
#include "JNI_WebRTC.h"

namespace jni
{
	namespace RTCIceServer
	{
		JavaLocalRef<jobject> toJava(JNIEnv * env, const webrtc::PeerConnectionInterface::IceServer & server)
		{
			const auto javaClass = JavaClasses::get<JavaRTCIceServerClass>(env);
			const auto & urls = server.urls;
			const auto & alpn = server.tls_alpn_protocols;
			const auto & ecv = server.tls_elliptic_curves;

			JavaArrayList urlList(env, urls.size());
			JavaArrayList alpnList(env, alpn.size());
			JavaArrayList ecList(env, ecv.size());

			for (auto & url : urls) {
				urlList.add(JavaString::toJava(env, url));
			}
			for (auto & protocol : alpn) {
				alpnList.add(JavaString::toJava(env, protocol));
			}
			for (auto & ec : ecv) {
				ecList.add(JavaString::toJava(env, ec));
			}

			auto username = JavaString::toJava(env, server.username);
			auto password = JavaString::toJava(env, server.password);
			auto hostname = JavaString::toJava(env, server.hostname);
			auto certPolicy = JavaEnums::toJava(env, server.tls_cert_policy);

			jobject iceServer = env->NewObject(javaClass->cls, javaClass->ctor);

			env->SetObjectField(iceServer, javaClass->urls, urlList.listObject());
			env->SetObjectField(iceServer, javaClass->tlsAlpnProtocols, alpnList.listObject());
			env->SetObjectField(iceServer, javaClass->tlsEllipticCurves, ecList.listObject());
			env->SetObjectField(iceServer, javaClass->username, username.get());
			env->SetObjectField(iceServer, javaClass->password, password.get());
			env->SetObjectField(iceServer, javaClass->hostname, hostname.get());
			env->SetObjectField(iceServer, javaClass->tlsCertPolicy, certPolicy.get());

			return JavaLocalRef<jobject>(env, iceServer);
		}

		webrtc::PeerConnectionInterface::IceServer toNative(JNIEnv * env, const JavaRef<jobject> & javaType)
		{
			const auto javaClass = JavaClasses::get<JavaRTCIceServerClass>(env);

			JavaObject obj(env, javaType);

			JavaLocalRef<jobject> urls = obj.getObject(javaClass->urls);
			JavaLocalRef<jstring> un = obj.getString(javaClass->username);
			JavaLocalRef<jstring> pw = obj.getString(javaClass->password);
			JavaLocalRef<jobject> cp = obj.getObject(javaClass->tlsCertPolicy);
			JavaLocalRef<jstring> hn = obj.getString(javaClass->hostname);
			JavaLocalRef<jobject> ap = obj.getObject(javaClass->tlsAlpnProtocols);
			JavaLocalRef<jobject> ec = obj.getObject(javaClass->tlsEllipticCurves);

			webrtc::PeerConnectionInterface::IceServer server;

			server.hostname = JavaString::toNative(env, hn);
			server.username = JavaString::toNative(env, un);
			server.password = JavaString::toNative(env, pw);

			server.urls = JavaList::toStringVector(env, urls);
			server.tls_alpn_protocols = JavaList::toStringVector(env, ap);
			server.tls_elliptic_curves = JavaList::toStringVector(env, ec);

			if (cp.get()) {
				server.tls_cert_policy = JavaEnums::toNative<webrtc::PeerConnectionInterface::TlsCertPolicy>(env, cp);
			}

			return server;
		}

		JavaRTCIceServerClass::JavaRTCIceServerClass(JNIEnv * env)
		{
			cls = FindClass(env, PKG"RTCIceServer");

			ctor = GetMethod(env, cls, "<init>", "()V");

			urls = GetFieldID(env, cls, "urls", LIST_SIG);
			username = GetFieldID(env, cls, "username", STRING_SIG);
			password = GetFieldID(env, cls, "password", STRING_SIG);
			tlsCertPolicy = GetFieldID(env, cls, "tlsCertPolicy", "L" PKG "TlsCertPolicy;");
			hostname = GetFieldID(env, cls, "hostname", STRING_SIG);
			tlsAlpnProtocols = GetFieldID(env, cls, "tlsAlpnProtocols", LIST_SIG);
			tlsEllipticCurves = GetFieldID(env, cls, "tlsEllipticCurves", LIST_SIG);
		}
	}
}